File: 数组/81. Search in Rotated Sorted Array II/binary_search.cpp
class Solution {
public:
    bool search(vector<int>& nums, int target) {
        // 时间复杂度：平均O(logn)，最差O(n)，因为如果不知道翻转点在哪侧，则可能遍历全部元素
        // 空间复杂度：O(1)
        if(!nums.size()) return false;
        int left = 0, right = nums.size()-1, mid;
        while(left < right) {
            mid = (left+right)/2;
            if(nums[mid]==target) return true;
            if(nums[right] > nums[left]) {
                if(nums[mid] > target) right = mid;
                else left = mid+1;
            } else {
                if(nums[mid]==nums[left] && nums[mid]==nums[right]) {
                    //不知道翻转点在哪
                    for(int i = left; i <= right; i++) {
                        if(nums[i] == target) return true;
                    }
                    return false;
                } else if(nums[mid] >= nums[left]) {
                    // 翻转点在右侧
                    if(nums[mid] < target) left = mid+1;
                    else if(nums[left] > target) left = mid+1;
                    else right = mid;
                } else {
                    // 翻转点在左侧
                    if(nums[mid] > target) right = mid;
                    else if(nums[right] < target) right = mid;
                    else left = mid+1;
                }
            }
        }
        return nums[left]==target;
    }
};

File: 查找/二分查找/315. Count of Smaller Numbers After Self/binary_search.py
# 时间复杂度 O(nlogn)
# 空间复杂度 O(n)
class Solution:
    def countSmaller(self, nums: List[int]) -> List[int]:
        if not len(nums):
            return []
        arr = [nums[-1]]
        res = [0]
        for val in nums[-2::-1]:
            count = -1
            if len(arr) == 1:
                if val > arr[0]:
                    arr.insert(0, val)
                    count = 1
                else:
                    arr.append(val)
                    count = 0
            else:
                # 二分查找
                left,right = 0,len(arr)-1
                while count == -1:
                    mid = int((left+right)/2)
                    if val > arr[mid]:
                        if mid == 0:
                            count = len(arr)
                            arr.insert(0, val)
                        elif val <= arr[mid-1]:
                            count = len(arr)-mid
                            arr.insert(mid, val)
                        else:
                            right = mid
                    else:
                        if mid == len(arr)-1:
                            count = 0
                            arr.append(val)
                        if val > arr[mid+1]:
                            count = len(arr)-mid-1
                            arr.insert(mid+1, val)
                        else:
                            left = mid+1
            res.insert(0,count)
        return res
                    